integration_test/list_test.go

```diff
@@ -42,3 +47,260 @@ func TestKrewList(t *testing.T) {
 	}
 	// TODO(ahmetb): install multiple plugins and see if the output is sorted
 }
+
+func TestKrewListJSON(t *testing.T) {
+	skipShort(t)
+
+	test, cleanup := NewTest(t)
+	defer cleanup()
+
+	test.Krew("install",
+		"--manifest", filepath.Join("testdata", "foo.yaml"),
+		"--archive", filepath.Join("testdata", "foo.tar.gz")).
+		RunOrFail()
+
+	expected := []byte("[\n    {\n        \"Name\": \"foo\",\n        \"Version\": \"v0.1.0\"\n    }\n]\n")
+
+	eventualList := test.WithIndex().Krew("list", "-o", "json").RunOrFailOutput()
+	if diff := cmp.Diff(eventualList, expected); diff != "" {
+		t.Fatalf("'list' output doesn't match:\n%s", diff)
+	}
+
+	var Plugin []struct {
+		Name    string
+		Version string
+	}
+
+	err := json.Unmarshal(eventualList, &Plugin)
+	if err != nil || Plugin[0].Name != "foo" || Plugin[0].Version != "v0.1.0" {
+		t.Fatalf("Error unmarshaling: %s. Plugin: \"%s\". Version: \"%s\".", err, Plugin[0].Name, Plugin[0].Version)
+	}
+}
+
+func TestKrewListJSONMulti(t *testing.T) {
+	skipShort(t)
+	test, cleanup := NewTest(t)
+	defer cleanup()
+
+	test.Krew("install", validPlugin).RunOrFail()
+	test.Krew("install",
+		"--manifest", filepath.Join("testdata", "foo.yaml"),
+		"--archive", filepath.Join("testdata", "foo.tar.gz")).
+		RunOrFail()
+
+	eventualList := test.Krew("list", "-o", "json").RunOrFailOutput()
```

Comment:
-o=json output should NOT be an `[]` json array.

PTAL at `-o=json` for pods etc. it's a json stream.

Reply:
So with the latest stable `kubectl` I get for `kubectl get pod -ojson`:
```json
{
    "apiVersion": "v1",
    "items": [],
    "kind": "List",
    "metadata": {
        "resourceVersion": "",
        "selfLink": ""
    }
}
```
when the namespace is empty. And if there are pods, they are listed in the `items` field. This is _not_ a json stream, right? However, we should also have the wrapping list object.

However, I totally agree that a list of json objects like a top-level structure as `[{..}, {..}]` is completely wrong.

Reply:
Actually you're right I missed that. It would be good to stick to that format.

Reply:
https://github.com/kubernetes-sigs/krew/pull/365#pullrequestreview-312725338

An objects list was what was discussed previously. I can change it to an object with the list contained within it but that'd require some relatively significant changes.

Reply:
Output currently looks like this for -o json.

```json
[
    {
        "Name": "grep",
        "Version": "v1.2.0"
    },
    {
        "Name": "ingress-nginx",
        "Version": "v0.25.0"
    },
    {
        "Name": "krew",
        "Version": "v0.3.1"
    },
    {
        "Name": "view-utilization",
        "Version": "v0.2.2"
    }
]
```

-o yaml

```yaml
- Name: grep
  Version: v1.2.0
- Name: ingress-nginx
  Version: v0.25.0
- Name: krew
  Version: v0.3.1
- Name: view-utilization
  Version: v0.2.2
```

Reply:
We can't really invent a new type for plugins.
We already have an `index.Plugin` type. I can't get behind introducing a format that is not going to be extensible and incompatible with the existing schema we have.

Reply:
Also you don't need to use metav1.List whatever.
You can just fake it.
I think we might need to perhaps start over and not use genericclioptions for this.

Reply:
I don't use metav1.List?

Reply:
But that's what's suggested above.
If we merge Krew into kubectl (a likely outcome) we'll need to adhere the standard format, so we can't take a change like this which we'd only have to break later on.